### ISSUE TITLE
Consumer throttling

### ISSUE THREAD

**Issue by PhilHannent** [weyoss/redis-smq#69]
This project is very interesting and straight forward to use.

One use-case I'd like is to throttle message consumption for the purpose of consuming 3rd party API calls. It would be good to have an configuration for the consumer which allows you to set the interval and quantity within that interval, much like this library:
https://www.npmjs.com/package/p-throttle

Is this something you would be interested in adding?

**Comment by PhantomRay** [comment]
Thought about that as well. If a queue has too many messages, dumping everything to consumer at once may not be ideal.

**Comment by weyoss** [comment]
@PhilHannent Thank you for opening this issue and for sharing your suggestion. 

> If a queue has too many messages, dumping everything to consumer at once may not be ideal.

@PhantomRay RedisSMQ does not dump everything  at once to a consumer. A consumer can receive only one message at a time. Once the message has been (un-)acknowledged, a new message may be received.

Returning to your question, I want to point out that the main goal of this project is to provide a simple message queue, usable, and with focus on message processing performance. 

If a feature may impact overall performance or may require a lot of changes and refactoring, it would not be promoted for implementation.

> One use-case I'd like is to throttle message consumption for the purpose of consuming 3rd party API calls. It would be good to have an configuration for the consumer which allows you to set the interval and quantity within that interval

A queue may have many consumers. So implementing message throttling at the consumer layer would not be a good idea. An ideal alternative sounds like setting a rate limit for delivering messages from a given queue. 

I am not going to draw any early conclusions for now. Let me think about it for a while.

I will keep you updated.

**Comment by PhantomRay** [comment]
Thank you for the explanation. 

My use case is a client will set up a persistent connection using server send event (SSE) connection or web socket connection. Then the sever will use consumer to consume the messages. On the server side, the consumer may consume messages quickly enough and dump them over the persistent connection, but the client may not be able to handle them quickly enough.

**Comment by weyoss** [comment]
This feature will be implemented in v6.2.0

**Comment by PhantomRay** [comment]
super awesome!

**Comment by weyoss** [comment]
This feature has been implemented in v6.2.0. 

See https://github.com/weyoss/redis-smq/blob/master/docs/queue-rate-limiting.md

**Comment by weyoss** [comment]
Closing as resolved.

**Comment by weyoss** [comment]
Recently a bug has been fixed in `redis-smq@6.2.2` regarding queue rate limiting. If you are using an older `redis-smq` version please upgrade.

The bug occurred only if you provided the queue parameter as a string and caused saving invalid Redis keys. For example:

```javascript
  queueManager.setQueueRateLimit('notifications', { limit: 200, interval: 60000 }, (err) => {
    // ...
  })
```
In the example the queue parameter was provided as a string (notifications), implying to use the default namespace, but in fact the bug caused no namespace to be used and saved an invalid Redis key.

`redis-smq@6.2.2` has fixed this issue.
